# Step 1: Import the dataset

In [53]:
import pandas as pd


In [108]:
netflix_original  = pd.read_csv("netflix_titles.csv")

In [109]:
netflix = netflix_original.copy()    #creating a copy of original data

In [110]:
netflix.info()   # shows column names, data types, and how many values are missing

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [111]:
netflix.head()    # shows the first 5 rows ("head" = top part)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [112]:
netflix.tail() #last 5 rows of data

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
8802,s8803,Movie,Zodiac,David Fincher,"Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...",United States,"November 20, 2019",2007,R,158 min,"Cult Movies, Dramas, Thrillers","A political cartoonist, a crime reporter and a..."
8803,s8804,TV Show,Zombie Dumb,NaN,NaN,NaN,"July 1, 2019",2018,TV-Y7,2 Seasons,"Kids' TV, Korean TV Shows, TV Comedies","While living alone in a spooky town, a young g..."
8804,s8805,Movie,Zombieland,Ruben Fleischer,"Jesse Eisenberg, Woody Harrelson, Emma Stone, ...",United States,"November 1, 2019",2009,R,88 min,"Comedies, Horror Movies",Looking to survive in a world taken over by zo...
8805,s8806,Movie,Zoom,Peter Hewitt,"Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...",United States,"January 11, 2020",2006,PG,88 min,"Children & Family Movies, Comedies","Dragged from civilian life, a former superhero..."
8806,s8807,Movie,Zubaan,Mozez Singh,"Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...",India,"March 2, 2019",2015,TV-14,111 min,"Dramas, International Movies, Music & Musicals",A scrappy but poor boy worms his way into a ty...


In [113]:
netflix.shape       # (rows, columns) — the "dimensions" (size) of the table

(8807, 12)

In [114]:
netflix.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

# Insights
The dataset comprises 8,807 rows and 12 columns.
The movie release years in the dataset range from a minimum of 1925 to a maximum of 2021, indicating that it covers movies from 1925 to 2021.

# Step 2: Identify and handle missing values

In [115]:
# Identify: check how many missing values each column has
netflix.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [116]:
netflix['country'] = netflix['country'].fillna('Unknown')   # fillna = fill missing values with something
netflix['cast'] = netflix['cast'].fillna('Unknown')


In [117]:
netflix['rating'] = netflix['rating'].fillna(netflix['rating'].mode()[0])  # mode = most frequent value

In [118]:
#  If a column is mostly empty and not useful, drop it
netflix = netflix.drop(columns=['director'])                 # drop = remove


In [119]:
#remove rows that have missing values in critical columns  
#since dates are unique events, not repeating categories mode not work on this  
netflix = netflix.dropna(subset=['date_added', 'duration'])

# Step 3: Remove duplicate records and formatting inconsistencies

In [120]:
# Duplicate = an exact repeated row
netflix = netflix.drop_duplicates()



In [121]:
# Formatting inconsistency = same value written in different styles (e.g., "usa", "USA", "U.S.A")
netflix['country'] = netflix['country'].str.strip()        # strip = remove extra spaces
netflix['country'] = netflix['country'].str.title()         # title case = "United States" (capitalizes each word)


# Step 4: Standardize columns such as Country, Rating, and Type

In [122]:
# Check if "USA" , appears anywhere in the column
netflix['country'].str.contains('USA', na=False).sum()


0

In [98]:
 # Check if "United States of America" , appears anywhere in the column
netflix['country'].str.contains('United States of America', na=False).sum()

0

In [104]:
netflix['rating'].unique()     # there is no problem in standardize of rating 

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', 'NR', 'TV-Y7-FV', 'UR'], dtype=object)

In [124]:
netflix['type'].unique() # there is no problem in standardize of type

array(['Movie', 'TV Show'], dtype=object)

# Step 5: Export the cleaned dataset

In [126]:
netflix.to_csv("netflix_cleaned.csv", index=False)